wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,106 @@
  * Add/update/retrieve Tapestry post and its child nodes
  * 
  */
-class TapestryController {}
\ No newline at end of file
+
+class TapestryController {
+    const POST_TYPES = [
+        'TAPESTRY' => 'tapestry',
+        'TAPESTRY_NODE' => 'tapestry_node'
+    ];
+    const ERRORS = [
+        'INVALID_POST_ID' => [
+            'MESSAGE' => 'PostID is invalid',
+            'STATUS' => ['status' => 404]
+        ]
+    ];
+
+    /**
+     * Update Tapestry nodes first then
+     * Update the existing Tapestry if the postId is provided
+     * Otherwise, a new Tapestry will be created
+     * 
+     * @param  Object @tapestry The Tapestry data
+     * @param  Number @postId The postId of the Tapestry
+     * @return Object @tapestry
+     */
+    public function updateTapestry($tapestry, $postId = null) {
+        // TODO: use isValidPostID() utlility function
+        if (!isset($postId)) {
+            $postId = $this->updateWordPressPost($tapestry, 'tapestry');
+        }
+        $this->updateNodes($tapestry->nodes, $postId);
```

Comment:
Please add an empty line between each piece of functionality in your code. Essentially, you'd be grouping your code so it's more clear, even if you are not commenting what's happening in each part of your function. Right now, you are adding empty lines between some of your code, but it's a bit arbitrary and doesn't serve any purpose or make the code easier to read.